BLOG: Add post about dataframe interoperability (PyCon Lithuania 2024 talk) (#841)
Quansight/Quansight-website: apps/labs/posts/dataframe-interop-pycon-lit-2024.md

---
title: 'Dataframe interoperability - what has been achieved, and what comes next?'
authors: [marco-gorelli]

> published: April 9, 2024

Reviewer: ```suggestion
published: May 21, 2024
```

description: 'An overview of the dataframe landscape, and solution to the "we only support pandas" problem'
category: [PyData ecosystem]
featuredImage:
  src: /posts/dataframe-interop-pycon-lit-2024/narwhals_small.jpg
  alt: 'Narwhals logo'
hero:
  imageSrc: /posts/dataframe-interop-pycon-lit-2024/narwhals_medium.jpg
  imageAlt: 'Narwhals logo'
---

# DataFrame Interoperability

> I attended PyConLithuania 2024, and had a blast! I don't speak Lithuanian, and probably

Reviewer: ```suggestion
I attended PyCon Lithuania 2024, and had a blast! I don't speak Lithuanian, and probably
```

neither did half the attendees. But - so long as we all stuck to a simple and clear subset
of the English language - we could all understand each other and exchange ideas. I actually
wanted to learn some Lithuanian in preparation for the event, but Duolingo didn't offer such
a course, and as a millennial, I was out of options.

Dataframes, like languages, have their similarities and differences, and it's easy to
make mistakes. If you learn a bit of
Spanish, you might think that "estoy embarazado" means "I'm embarrassed", but it actually
means "I'm pregnant". Similarly, after learning a bit of pandas and Polars, you might
expect
```python
import pandas as pd
import polars as pd

print((3 in pd.Series([1,2,3])) == (3 in pl.Series([1,2,3])))
```

> to print `'True'` - but no so! pandas checks if `3` is in the index, whereas Polars checks

Reviewer: ```suggestion
to print `'True'` - but not so! pandas checks if `3` is in the index, whereas Polars checks
```

if `3` is in the values.

This little prelude was just to establish two premises:

- writing dataframe-agnostic code is hard!
- a simple and clear common language can enable collaboration

I'll tell you about the dataframe landscape, the interchange protocol, how to write
dataframe-agnostic code, and where we go from here.

> ## pandas are everywhere

Reviewer: ```suggestion
## pandas is everywhere
```

There's a great array of diverse dataframes out there. And the way that data science libraries

> have typically to such diversity is to support pandas...

Reviewer: ```suggestion
typically have to such diversity is to support only pandas...
```

![Image showing a pumpkin being fed to someone through a funnel. The input to the funnel is labelled
"Dataframe X", and the output "Dataframe Y"](/posts/dataframe-interop-pycon-lit-2024/interchange.png)

Let's look at how to use it - here's how you can convert any dataframe to pandas
(so long as it implements the interchange protocol):

```python
import pandas as pd

df_pandas = pd.api.interchange.from_dataframe(df_any)
```

Similarly, to convert to Polars:

```python
import polars as pl

df_polars = pl.from_dataframe(df_any)
```

Note that although the `from_dataframe` function is standardised,
where it appears in the API isn't, so there's no completely agnostic way
of round-tripping back to your starting dataframe class.

Nonetheless, does it work? Is it reliable?

- Converting to pandas: reliable enough after pandas 2.0.2, though may sometimes
  raise unnecessarily in some versions. It's already used by the `seaborn` and
  `plotly` dataframe libraries.
- Converting from pandas: unreliable, don't use it. There are cases when the results
  are literal nonsense. These have generally been fixed, and will be available in

>   pandas 3.0+. For now, however, if you're using the interchange protocol to convert

Reviewer: ```suggestion
 pandas 3.0. For now, however, if you're using the interchange protocol to convert
```

interchange protocol partially addresses the first one (so long as the user doesn't
expect their computation to round-trip back to their original class).

Addressing all four points seems ambitious - is it even possible?

## And now for something completely different: Narwhals

Look at how happy this panda and polar bear looking, chilling out with their narwhal friend:

![Image showing a panda, a polar bear, and a narwhal chilling out in an office together](/posts/dataframe-interop-pycon-lit-2024/narwhals_small.jpg)

As you may have guessed from the image, Narwhals aims to bring pandas and Polars together.

The way you write dataframe-agnostic code using Narwhals is:

1. Use `narwhals.from_native` to wrap any dataframe in a Narwhals class.
2. Use the subset of the Polars API supported by Narwhals.
3. Use `narwhals.to_native` to return the object to the user in its original class.

If you follow these steps, then congrats, your code will work seamlessly across pandas,
Polars, Modin, and cuDF. But most importantly, GPU code can stay on GPU, and lazy code
can stay lazy!

To see a complete example of Narwhals in action, its website features

> [a little tutorial](https://marcogorelli.github.io/narwhals/basics/complete_example/).

Reviewer: this link is broken, as are the other ones to this site

Regarding the four problems of libraries only supporting pandas, Narwhals addresses all
of them:

- Users need not do any data conversion manually (in fact, neither does Narwhals, it only
  translates syntax).
- Data which starts on GPU can stay on GPU.
- Data which starts lazy can stay lazy.
- No heavy dependencies necessary (as we'll soon see, Narwhals is about as light as it gets).

Finally, [Narwhals is also extensible](https://marcogorelli.github.io/narwhals/extending/),

> meaning that libraries can become compatible without having to asking any other library

Reviewer: ```suggestion
meaning that libraries can become compatible without having to ask any other library
```

to change anything.

### Did you get deja-vu? Does this remind you of Ibis?

A related project is Ibis, which [presents itself as a dataframe standard](https://voltrondata.com/resources/open-source-standards),
and dispatches to 20+ backends. Some differences with Narwhals are:

- Narwhals is ~1000 times lighter
- Narwhals only supports 4 backends, Ibis more than 20
- Narwhals is new, whereas Ibis is mature and production-ready

The first point is a major differentiator:
if your objective is to make your dataframe library dataframe-agnostic, and package size is a concern to you,

> then Narwhals is currently (as of April 2024) a more realistic solution - Ibis alone (310 MB) would put you above

Reviewer: "Ibis alone" should probably state that it's Ibis plus all its required dependencies?

the AWS Lambda package size limit (250MB).

> ![Image relative sizes of Narwhals (0.3 MB) vs Ibis (310 MB)](/posts/dataframe-interop-pycon-lit-2024/narwhals_vs_ibis.png)

Reviewer: This image is very rough, I suggest to remove it.

> Note that the projects are not in competition and have different goals.

Reviewer: This leaves the reader wondering quite a bit. Maybe add one sentence for each?
